File: backend/app/domains/expedicao/expedicao_model.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database.conexao import Base
from app.shared.sync_mixin import IdMixin, SyncMixin

# Nenhum model de `pedidos` ou `usuarios` é importado aqui — as FKs abaixo
# apontam pra tabelas de outros domínios só pelo nome, e é o banco (FK real)
# quem recusa um id inexistente. Ver ARCHITECTURE.md → "Validação de id por
# foreign key". `pedido_itens` é documento (não cadastro), então nem
# `relationship()` é criado pra ele a partir daqui — quem precisa exibir
# dado do pedido consulta o domínio de pedidos separadamente.


class ExpedicaoPedidoStatus(Base, IdMixin, SyncMixin):
    """
    Em que ponto do galpão cada pedido está. Uma linha por pedido, atualizada
    conforme as etapas avançam.

    Existe em vez de gravar direto em `pedidos.status_id` porque aquele campo é
    da integração: um PUT do ERP no mesmo pedido sobrescreveria o andamento da
    expedição. Aqui o dono da escrita é só este domínio.

    `status_id` aponta para o mesmo catálogo `pedido_status` usado pelo pedido —
    o vocabulário de status é um só no sistema, não dois.
    """

    __tablename__ = "expedicao_pedido_status"

    # Único: é o status ATUAL, não histórico. Quem quiser reconstruir a linha do
    # tempo tem as datas de início e fim dos processos e dos itens.
    pedido_id: Mapped[str] = mapped_column(
        ForeignKey("pedidos.id"), nullable=False, unique=True, index=True
    )
    status_id: Mapped[str] = mapped_column(ForeignKey("pedido_status.id"), nullable=False, index=True)


class ExpedicaoAtribuicao(Base, IdMixin, SyncMixin):
    """
    Quem é o responsável por uma etapa de um pedido.

    É por etapa, não por pedido: o caso normal do galpão é uma pessoa separar e
    outra conferir o mesmo pedido, então `(pedido_id, tipo)` é que identifica a
    designação — não `pedido_id` sozinho.

    Esta tabela é o que decide o que cada um ENXERGA na listagem: operador sem
    a permissão `expedicao.atribuir` só vê pedido em que ele é o responsável de
    alguma etapa (ver `listar_pedidos` em expedicao_service.py). Por isso ela
    não é um enfeite de UI — é regra de acesso, e mora no backend.

    Desatribuir é `marcar_apagado`, nunca DELETE: quem atribuiu, para quem e
    quando é rastro de auditoria que o galpão vai querer no dia em que um
    pedido sair errado.
    """

    __tablename__ = "expedicao_atribuicoes"

    pedido_id: Mapped[str] = mapped_column(ForeignKey("pedidos.id"), nullable=False, index=True)
    # 'separacao' | 'conferencia' — mesmo par do TipoProcesso do contrato.
    tipo: Mapped[str] = mapped_column(String(20), nullable=False, index=True)
    usuario_id: Mapped[str] = mapped_column(ForeignKey("usuarios.id"), nullable=False, index=True)
    # Quem distribuiu. Não é o mesmo que `usuario_id`, e é o que responde
    # "quem mandou fulano separar esse pedido?".
    atribuido_por_id: Mapped[str] = mapped_column(ForeignKey("usuarios.id"), nullable=False, index=True)
    data_atribuicao: Mapped[datetime] = mapped_column(DateTime(), nullable=False)

    # Sem UniqueConstraint em (pedido_id, tipo): o soft delete deixaria linhas
    # apagadas ocupando a chave, e o MySQL não tem índice único parcial. Quem
    # garante um responsável vivo por etapa é `atribuir` no service, que apaga
    # a atribuição anterior antes de gravar a nova.


class Separacao(Base, IdMixin, SyncMixin):
    __tablename__ = "expedicao_separacoes"

    pedido_id: Mapped[str] = mapped_column(ForeignKey("pedidos.id"), nullable=False, index=True)
    # De quem é o TRABALHO. Continua sendo o operador mesmo quando quem clicou
    # foi o gerente — é por isso que o par de colunas abaixo existe.
    usuario_inicio_id: Mapped[str] = mapped_column(ForeignKey("usuarios.id"), nullable=False, index=True)
    usuario_fim_id: Mapped[str | None] = mapped_column(ForeignKey("usuarios.id"), nullable=True, index=True)
    # Quem CLICOU, quando não foi o próprio operador. NULL = o operador abriu e
    # fechou sozinho, que é o caso normal.
    #
    # O galpão nem sempre tem coletor para todo mundo: o gerente despacha uma
    # pessoa para separar no papel, e depois registra o início e o fim aqui. Sem
    # estas duas colunas, o sistema diria que o gerente separou o pedido (falso)
    # ou que o operador operou o sistema (também falso) — e o relatório de
    # produtividade sairia errado nos dois casos.
    usuario_gestor_inicio_id: Mapped[str | None] = mapped_column(
        ForeignKey("usuarios.id"), nullable=True, index=True
    )
    usuario_gestor_fim_id: Mapped[str | None] = mapped_column(
        ForeignKey("usuarios.id"), nullable=True, index=True
    )
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="em_andamento")
    data_inicio: Mapped[datetime | None] = mapped_column(DateTime(), nullable=True, default=None)
    # Quando a PRIMEIRA leitura foi registrada. Diferente de `data_inicio`, que
    # é quando o processo foi aberto: entre abrir a lista e bipar o primeiro
    # item pode passar muito tempo (o operador ainda vai até o endereço). O
    # tempo de trabalho de verdade se mede daqui, não da abertura.
    data_primeiro_bipe: Mapped[datetime | None] = mapped_column(
        DateTime(), nullable=True, default=None
    )
    data_fim: Mapped[datetime | None] = mapped_column(DateTime(), nullable=True, default=None)

    itens: Mapped[list["SeparacaoItem"]] = relationship(
        back_populates="separacao",
        cascade="all, delete-orphan",
        lazy="selectin",
        order_by="SeparacaoItem.sync_created_at",
    )


class SeparacaoItem(Base, IdMixin, SyncMixin):
    """
    Não tem coluna de status: o estado do item é derivado das datas.
    `data_inicio` nula = pendente; início preenchido sem fim = em andamento;
    `data_fim` preenchida = finalizado. Uma coluna a menos, e zero chance de
    um campo `status` divergir das datas que medem o tempo gasto por item.
    """

    __tablename__ = "expedicao_separacao_itens"

    separacao_id: Mapped[str] = mapped_column(
        ForeignKey("expedicao_separacoes.id", ondelete="CASCADE"), nullable=False, index=True
    )
    pedido_item_id: Mapped[str] = mapped_column(ForeignKey("pedido_itens.id"), nullable=False, index=True)
    quantidade_separada: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    data_inicio: Mapped[datetime | None] = mapped_column(DateTime(), nullable=True, default=None)
    data_fim: Mapped[datetime | None] = mapped_column(DateTime(), nullable=True, default=None)
    divergente: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    # Gerente que liberou a finalização com quantidade abaixo da pedida.
    # NULL quando o item fechou completo (o caso normal). Sem essa coluna,
    # "quem autorizou a falta" não fica rastreável em lugar nenhum.
    usuario_autorizador_id: Mapped[str | None] = mapped_column(
        ForeignKey("usuarios.id"), nullable=True, index=True
    )

    separacao: Mapped["Separacao"] = relationship(back_populates="itens")


class Conferencia(Base, IdMixin, SyncMixin):
    __tablename__ = "expedicao_conferencias"

    pedido_id: Mapped[str] = mapped_column(ForeignKey("pedidos.id"), nullable=False, index=True)
    usuario_inicio_id: Mapped[str] = mapped_column(ForeignKey("usuarios.id"), nullable=False, index=True)
    usuario_fim_id: Mapped[str | None] = mapped_column(ForeignKey("usuarios.id"), nullable=True, index=True)
    # Mesmo papel do par em Separacao — ver lá.
    usuario_gestor_inicio_id: Mapped[str | None] = mapped_column(
        ForeignKey("usuarios.id"), nullable=True, index=True
    )
    usuario_gestor_fim_id: Mapped[str | None] = mapped_column(
        ForeignKey("usuarios.id"), nullable=True, index=True
    )
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="em_andamento")
    data_inicio: Mapped[datetime | None] = mapped_column(DateTime(), nullable=True, default=None)
    # Quando a PRIMEIRA leitura foi registrada. Diferente de `data_inicio`, que
    # é quando o processo foi aberto: entre abrir a lista e bipar o primeiro
    # item pode passar muito tempo (o operador ainda vai até o endereço). O
    # tempo de trabalho de verdade se mede daqui, não da abertura.
    data_primeiro_bipe: Mapped[datetime | None] = mapped_column(
        DateTime(), nullable=True, default=None
    )
    data_fim: Mapped[datetime | None] = mapped_column(DateTime(), nullable=True, default=None)

    # --- O desfecho da baixa no ERP (GESTCOM) ---
    #
    # A conferência fechar aqui e o pedido fechar lá são duas coisas: o Oracle
    # é outro banco, outra transação, e pode estar fora do ar no exato momento
    # em que o operador termina de bipar. Sem estas duas colunas o pedido ficava
    # "conferido" no ELLOTEC e `PED` no ERP, sem ninguém saber por quê.
    #
    # `finalizado_origem_em` é coluna de negócio própria, e não `sync_*`: o
    # instante em que o ERP aceitou a baixa é um FATO, e campo de sincronização
    # nunca entra na regra de negócio (ver ARCHITECTURE.md).
    finalizado_origem_em: Mapped[datetime | None] = mapped_column(
        DateTime(), nullable=True, default=None
    )
    # Por que a última tentativa foi recusada. Sobrescrito a cada tentativa e
    # limpo quando dá certo — é o que a tela mostra para o operador entender se
    # o caso é "tenta de novo" ou "chama o faturamento".
    motivo_falha_origem: Mapped[str | None] = mapped_column(String(255), nullable=True, default=None)
    # Quem tentou e quando, na ÚLTIMA tentativa — deu certo ou não.
    #
    # Sem este par, `motivo_falha_origem` conta o que aconteceu mas não em nome
    # de quem nem a que horas, e a pergunta que aparece depois é sempre essa:
    # "quem tentou fechar este pedido e não conseguiu?". Numa recusa por conta
    # sem vínculo no ERP, saber QUAL conta clicou é o que resolve o caso.
    #
    # Sobrescrito a cada tentativa, e NÃO limpo no sucesso: aí ele passa a
    # responder "quem fechou o pedido", junto com `finalizado_origem_em`.
    tentativa_origem_usuario_id: Mapped[str | None] = mapped_column(
        ForeignKey("usuarios.id"), nullable=True, index=True, default=None
    )
    # Coluna de negócio própria, e não `sync_updated_at`: a linha é tocada por
    # outras escritas, e a hora da tentativa é um fato que não pode se mexer
    # sozinho (ver ARCHITECTURE.md).
    tentativa_origem_em: Mapped[datetime | None] = mapped_column(
        DateTime(), nullable=True, default=None
    )

    itens: Mapped[list["ConferenciaItem"]] = relationship(
        back_populates="conferencia",
        cascade="all, delete-orphan",
        lazy="selectin",
        order_by="ConferenciaItem.sync_created_at",
    )


class ConferenciaItem(Base, IdMixin, SyncMixin):
    """Mesmo desenho de SeparacaoItem — estado derivado das datas."""

    __tablename__ = "expedicao_conferencia_itens"

    conferencia_id: Mapped[str] = mapped_column(
        ForeignKey("expedicao_conferencias.id", ondelete="CASCADE"), nullable=False, index=True
    )
    pedido_item_id: Mapped[str] = mapped_column(ForeignKey("pedido_itens.id"), nullable=False, index=True)
    quantidade_conferida: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    divergente: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    data_inicio: Mapped[datetime | None] = mapped_column(DateTime(), nullable=True, default=None)
    data_fim: Mapped[datetime | None] = mapped_column(DateTime(), nullable=True, default=None)
    # Mesmo papel de SeparacaoItem.usuario_autorizador_id — ver lá.
    usuario_autorizador_id: Mapped[str | None] = mapped_column(
        ForeignKey("usuarios.id"), nullable=True, index=True
    )

    conferencia: Mapped["Conferencia"] = relationship(back_populates="itens")

File: src/app/domains/expedicao/expedicao.model.ts
/** Separação e conferência têm o mesmo ciclo de vida — o tipo entra na URL. */
export type TipoProcesso = 'separacao' | 'conferencia';

/** Estado de um item dentro de um processo. Derivado das datas no backend,
 *  nunca calculado aqui (a regra é de lá, ver expedicao_service.py). */
export type SituacaoItem = 'pendente' | 'em_andamento' | 'finalizado';

export type StatusProcesso = 'nao_iniciada' | 'em_andamento' | 'finalizada';

export interface SituacaoProcesso {
  id: string | null;
  status: StatusProcesso;
  usuarioId: string | null;
  usuarioNome: string | null;
  itensFinalizados: number;
  itensTotal: number;
  temDivergencia: boolean;
  /** Quando a primeira leitura foi registrada — é daqui que o tempo de
   *  trabalho conta, não da abertura do processo. Nulo = ninguém bipou ainda. */
  dataPrimeiroBipe: string | null;
  /** Quando o processo fechou. Nulo = ainda em andamento. */
  dataFim: string | null;
  /** Quando a etapa foi aberta. Nulo = etapa nunca iniciada. */
  dataInicio: string | null;
  /** Execução delegada: quem CLICOU, quando não foi o próprio operador. O
   *  gerente registra o início e o fim da etapa no nome de quem está separando
   *  no papel, porque não há coletor para todo mundo. Nulos no caso normal. */
  usuarioGestorInicioNome: string | null;
  usuarioGestorFimNome: string | null;
  /** Derivado dos dois acima pelo backend — a tela não recalcula. */
  delegado: boolean;
  /** Quando o ERP aceitou a baixa do pedido (status FEC). Nulo = a conferência
   *  fechou aqui mas o pedido continua aberto lá. Só a conferência tem: não há
   *  o que fechar no sistema de origem ao terminar uma separação. */
  finalizadoOrigemEm: string | null;
  /** Por que a última tentativa de finalizar no ERP foi recusada. */
  motivoFalhaOrigem: string | null;
  /** Quem tentou e quando, na última tentativa — deu certo ou não. Responde
   *  "quem tentou fechar este pedido, e a que horas?" sem depender de ninguém
   *  lembrar da mensagem que viu na tela. */
  tentativaOrigemEm: string | null;
  tentativaOrigemUsuarioNome: string | null;
}

/**
 * De quando o relógio da etapa conta.
 *
 * No caminho normal é o primeiro bipe, e não a abertura: abrir a lista e andar
 * até o endereço não é tempo de separação.
 *
 * Na execução delegada ninguém bipa — o gerente registra o início, o operador
 * separa no papel e o gerente registra o fim. Ali `dataPrimeiroBipe` é nulo
 * para sempre, e cair no `dataInicio` é o que faz a coluna mostrar a hora e a
 * duração em vez de um traço numa etapa que começou e terminou.
 */
export function inicioDaEtapa(situacao: SituacaoProcesso): string | null {
  return situacao.dataPrimeiroBipe ?? situacao.dataInicio;
}

/** Duração já quebrada em partes, pronta para o badge. */
export interface Duracao {
  texto: string;
  emAndamento: boolean;
}

/**
 * "1d 2h 30m" a partir de dois instantes. `fim` nulo mede até agora — é o que
 * faz o badge de um processo aberto mostrar quanto tempo ele já está correndo.
 *
 * Minutos são a menor unidade de propósito: segundos numa listagem só poluem,
 * e ninguém decide nada no galpão por causa de 40 segundos.
 */
export function duracaoEntre(inicio: string | null, fim: string | null): Duracao | null {
  if (!inicio) return null;
  const emAndamento = !fim;
  const ms = (fim ? new Date(fim).getTime() : Date.now()) - new Date(inicio).getTime();
  if (Number.isNaN(ms) || ms < 0) return null;

  const minutos = Math.floor(ms / 60000);
  const dias = Math.floor(minutos / 1440);
  const horas = Math.floor((minutos % 1440) / 60);
  const restoMinutos = minutos % 60;

  const partes: string[] = [];
  if (dias) partes.push(`${dias}d`);
  if (horas) partes.push(`${horas}h`);
  // Sempre mostra minutos quando não há dia nem hora, senão "menos de 1 min"
  // apareceria como texto vazio.
  if (restoMinutos || !partes.length) partes.push(`${restoMinutos}m`);

  return { texto: partes.join(' '), emAndamento };
}

/** Responsável designado por uma etapa. Null na listagem = ninguém designado.
 *  É por isto que o operador comum enxerga (ou não) a linha — o filtro é do
 *  backend, aqui é só exibição. */
export interface Atribuicao {
  usuarioId: string;
  usuarioNome: string;
  atribuidoPorNome: string | null;
  dataAtribuicao: string;
}

/** Opção do seletor de responsável — só quem pode executar aquela etapa. */
export interface Operador {
  id: string;
  nome: string;
}

export interface PedidoExpedicaoLista {
  pedidoId: string;
  /** Nulo quando a origem externa ainda não deu número ao pedido. Use
  *  `numeroExibicao` para renderizar — ela cobre esse caso. */
  numero: string | null;
  sistemaOrigemId: string | null;
  dataPedido: string;
  /** Status vindo do ERP (PED, OK, CAN…). */
  statusPedido: string;
  /** Só pedido em PED pode abrir separação ou conferência — quem decide é o
   *  backend, a tela só reflete. */
  podeIniciar: boolean;
  /** Nulo = endereçamento em ordem. Preenchido = a primeira pendência do
   *  pedido (já com o código do produto na frente). Quando vem preenchido,
   *  `podeIniciar` é false — a tela não precisa combinar os dois. */
  bloqueioEnderecamento: string | null;
  /** Data da última alteração: é por ela que a listagem ordena e filtra. */
  alteradoEm: string;
  clienteNomeFantasia: string;
  clienteCnpj: string;
  clienteCidadeNome: string;
  clienteCidadeUf: string;
  /** Quando o pedido foi liberado no ERP (ex: aprovação de crédito). É deste
   *  marco que o ciclo da expedição é contado — não da data do pedido. */
  liberadoEm: string | null;
  /** Empresa emissora — matriz ou filial. */
  empresaId: string;
  empresaNome: string;
  /** Nome curto do dia a dia ("Matriz", "BSB") — o que a coluna "Emp" mostra.
   *  Nome fantasia inteiro não cabe numa coluna estreita, e numa fila de várias
   *  empresas o que se precisa é distinguir, não ler o nome completo. Nulo
   *  enquanto ninguém cadastrar o apelido da empresa. */
  empresaApelido: string | null;
  quantidadeItens: number;
  quantidadeTotal: number;
  separacao: SituacaoProcesso;
  conferencia: SituacaoProcesso;
  atribuicaoSeparacao: Atribuicao | null;
  atribuicaoConferencia: Atribuicao | null;
}

/** Formato de resposta do GET /expedicao/pedidos, paginado no banco: são
 *  centenas de milhares de pedidos, e trazer tudo estoura o navegador. */
export interface PedidoExpedicaoListaPaginada {
  items: PedidoExpedicaoLista[];
  total: number;
  page: number;
  perPage: number;
  sort: string;
  sortType: string;
  /** Quantos pedidos há em cada situação NO PERÍODO — uma chave por
   *  `FiltroSituacao`, inclusive 'todos'.
   *
   *  Contam só o período: ignoram termo, status, empresa, operador e a própria
   *  situação escolhida. São um painel fixo do galpão naquele intervalo, então
   *  não mudam enquanto a pessoa mexe nos filtros — é isso que permite usá-los
   *  para navegar ("tem 3 parados, vou ali ver"). */
  contagensPorSituacao: Record<FiltroSituacao, number>;
}

/** Colunas por onde o servidor sabe ordenar. Só existem as que saem da própria
 *  tabela `pedidos` — cidade, quantidade de itens e andamento das etapas vêm de
 *  outras tabelas e não entram na consulta paginada. */
export type ColunaOrdenavel =
  | 'numero'
  | 'data_pedido'
  | 'cliente_nome_fantasia'
  | 'quantidade_itens'
  | 'liberado_em'
  | 'sync_updated_at';

/** Um endereço em que o lote do item está, e quanto tem nele.
 *
 *  A quantidade é o ponto: antes a expedição mostrava só o total do item
 *  somado, o que não dizia ao operador quanto pegar em cada prateleira. */
export interface EnderecoItem {
  enderecoId: string;
  descricao: string;
  quantidade: number;
}

export interface ItemPedidoExpedicao {
  pedidoItemId: string;
  produtoId: string;
  produtoCodigo: string;
  produtoDescricao: string;
  /** Marca e códigos de barras vêm do cadastro vivo do produto (não do
   *  snapshot do pedido): é o que está impresso na caixa que o operador tem na
   *  mão. O da nota e os de logística vão separados porque a tela os rotula
   *  diferente. */
  produtoMarcaNome: string;
  produtoCodigoBarraNotas: string | null;
  produtoCodigosBarrasLogistica: string[];
  produtoDun14: string | null;
  /** Onde a mercadoria está guardada, com quanto tem em cada lugar. É LISTA
   *  porque um lote se espalha por vários endereços do galpão — vem de
   *  `estoque_endereco_lote` (domínio `enderecamento`), não da linha do
   *  pedido. */
  enderecos: EnderecoItem[];
  /** Soma das quantidades acima, calculada no backend. A tela NÃO soma por
   *  conta própria: é este número que decidiu o bloqueio, e refazer a conta em
   *  float no navegador daria um total diferente. */
  quantidadeEnderecada: number;
  /** Nulo = item consistente. Preenchido = a frase do quadro vermelho. */
  bloqueio: string | null;
  lote: string | null;
  quantidade: number;
  /** Unidades por embalagem de venda do produto (1 = vendido na unidade).
   *  Cada leitura no coletor vale essa quantidade. */
  quantidadeMultiplaVenda: number;
  separacaoSituacao: SituacaoItem;
  separacaoQuantidade: number;
  conferenciaSituacao: SituacaoItem;
  conferenciaQuantidade: number;
}

export interface PedidoExpedicaoDetalhe {
  pedidoId: string;
  numero: string | null;
  sistemaOrigemId: string | null;
  dataPedido: string;
  /** Status vindo do ERP. */
  statusPedido: string;
  /** Só pedido em PED autoriza abrir separação ou conferência. */
  podeIniciar: boolean;
  bloqueioEnderecamento: string | null;
  /** Só a barreira do ERP, sem o endereçamento. É ela que os botões de execução
   *  delegada olham: quem tem `expedicao.enderecamento.liberar` atravessa o
   *  endereçamento inconsistente, mas nunca o status errado. */
  statusPermiteIniciar: boolean;
  observacoes: string;
  vendedorNome: string | null;
  clienteCodigo: string | null;
  clienteRazaoSocial: string;
  clienteNomeFantasia: string;
  clienteCnpj: string;
  clienteEndereco: string;
  clienteBairro: string | null;
  clienteCep: string | null;
  clienteCidadeNome: string;
  clienteCidadeUf: string;
  quantidadeItens: number;
  quantidadeTotal: number;
  separacao: SituacaoProcesso;
  conferencia: SituacaoProcesso;
  /** Qual botão o rodapé deve oferecer. null = as duas etapas já fecharam. */
  proximaEtapa: TipoProcesso | null;
  atribuicaoSeparacao: Atribuicao | null;
  atribuicaoConferencia: Atribuicao | null;
  itens: ItemPedidoExpedicao[];
}

export interface ItemProcesso {
  pedidoItemId: string;
  produtoId: string;
  produtoCodigo: string;
  produtoDescricao: string;
  /** Unidade do cadastro do produto (UN, CX…). */
  produtoUnidade: string;
  /** Do cadastro vivo do produto — ver ItemPedidoExpedicao. */
  produtoMarcaNome: string;
  produtoCodigoBarraNotas: string | null;
  produtoCodigosBarrasLogistica: string[];
  produtoDun14: string | null;
  /** Ver ItemPedidoExpedicao — mesmos três campos, mesma origem. */
  enderecos: EnderecoItem[];
  quantidadeEnderecada: number;
  bloqueio: string | null;
  lote: string | null;
  quantidadePedida: number;
  quantidadeProcessada: number;
  /** Unidades por embalagem de venda do produto (1 = vendido na unidade). */
  quantidadeMultiplaVenda: number;
  dataInicio: string | null;
  dataFim: string | null;
  divergente: boolean;
  situacao: SituacaoItem;
}

export interface Processo {
  id: string;
  tipo: TipoProcesso;
  pedidoId: string;
  pedidoNumero: string;
  status: 'em_andamento' | 'finalizada';
  /** De quem é o TRABALHO — o operador. Não muda quando o gerente executa em
   *  nome dele. */
  usuarioInicioId: string;
  usuarioInicioNome: string | null;
  usuarioFimId: string | null;
  /** Quem CLICOU, quando não foi o próprio operador. Nulos no caso normal. */
  usuarioGestorInicioNome: string | null;
  usuarioGestorFimNome: string | null;
  dataInicio: string | null;
  dataFim: string | null;
  /** Mesmos quatro campos de SituacaoProcesso — ver lá. */
  finalizadoOrigemEm: string | null;
  motivoFalhaOrigem: string | null;
  tentativaOrigemEm: string | null;
  tentativaOrigemUsuarioNome: string | null;
  itens: ItemProcesso[];
}

/**
 * Os quatro números que o ERP pede para fechar o pedido. Não saem de lugar
 * nenhum do sistema: só existem depois de a mercadoria estar embalada, e é o
 * operador quem digita, no modal que abre quando a conferência termina.
 */
export interface FinalizacaoSistemaOrigem {
  /** Quantidade de volumes. Sempre inteiro — no ERP a coluna é texto
   *  (`VARCHAR2(10)`) e guarda dígitos, sem separador decimal. */
  volume: number;
  /** Espécie da embalagem — até 10 caracteres maiúsculos (CX, FD, SC, CAIXA…),
   *  que é o tamanho da coluna no ERP. */
  especie: string;
  /** Em quilos. */
  pesoLiquido: number;
  pesoBruto: number;
}

export interface CredencialGerente {
  usuarioGerente: string;
  senha: string;
}

export function rotuloTipo(tipo: TipoProcesso): string {
  return tipo === 'separacao' ? 'Separação' : 'Conferência';
}

/** Pedidos vindos de integração externa trazem o identificador de lá — quando
 *  presente, é ele que representa o pedido pra quem está olhando. Mesma regra
 *  de `numeroExibicaoPedido` no domínio de pedidos. */
export function numeroExibicao(pedido: {
  numero: string | null;
  sistemaOrigemId: string | null;
}): string {
  // O traço cobre o pedido externo que chegou sem número e ainda não recebeu um
  // — a tela precisa desenhar alguma coisa, e um espaço em branco pareceria bug.
  return pedido.sistemaOrigemId || pedido.numero || '—';
}

/** O item em andamento bloqueia todos os outros — é essa a regra que a tela de
 *  itens usa pra desabilitar os botões das outras linhas. */
export function itemEmAndamento(processo: Processo | null): ItemProcesso | undefined {
  return processo?.itens.find((item) => item.situacao === 'em_andamento');
}

/**
 * Situação do pedido no galpão — o filtro de uma linha da listagem.
 *
 * Mora no model, e não na tela, porque virou contrato de API: o valor vai na
 * query string de GET /expedicao/pedidos e é o backend que recorta a base
 * inteira por ele (ver SITUACOES_VALIDAS em expedicao_service.py). Enquanto o
 * filtro era aplicado sobre a página carregada, era detalhe da tela.
 */
export type FiltroSituacao =
  | 'todos'
  | 'nao_iniciados'
  | 'em_separacao'
  | 'aguardando_conferencia'
  | 'em_conferencia'
  | 'concluidos'
  | 'divergentes';

/**
 * Tudo que recorta a listagem. Um objeto e não onze parâmetros posicionais:
 * são todos opcionais entre si, e a ordem deles não significa nada para quem lê
 * a chamada.
 *
 * TODOS vão para o servidor. Nenhum filtra a lista já carregada — com ~230 mil
 * pedidos, o que chega na tela é uma amostra, e recortar a amostra responde
 * "não achei" para pedido que existe na página seguinte.
 */
export interface FiltrosPedidoExpedicao {
  page: number;
  perPage: number;
  dataInicio: string;
  dataFim: string;
  q: string;
  statusPedido: string[];
  /** Vazio = todas as empresas. */
  empresaId: string;
  /** Quem abriu a separação ou a conferência. Vazio = qualquer um. */
  operadorId: string;
  situacao: FiltroSituacao;
  sort: string;
  sortType: string;
}
